packages/@aws-cdk/assert/lib/canonicalize-assets.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * Reduce template to a normal form where asset references have been normalized
+ *
+ * This makes it possible to compare templates if all that's different between
+ * them is the hashes of the asset values.
+ *
+ * Currently only handles parameterized assets, but can (and should)
+ * be adapted to handle convention-mode assets as well when we start using
+ * more of those.
+ */
+export function canonicalizeTemplate(template: any): any {
```

Comment:
Can this operate on the assembly and then instead of regex search & replace actually *know* that there are assets in the stack? I think we should have an assembly available in all code paths leading to this.

Reply:
Maybe.

The code doesn't really become appreciably simpler if we do that: it still needs to do roughly the same things. As an example, we still write `This is the S3 bucket key for asset "fff6caa93ca53b583b89a8561e84b0f7"` in the template, so we're still going to have to regex replace for that specific substring.

What we'll save is only regex search, and I don't believe it's inaccurate or will lead to false positives. As written, it may lead to false negatives, but I don't think that's a big issue.

---

In terms of code changes required to do that:

A new API that takes that into account really only makes sense in terms of a `cxapi.CloudFormationStackArtifact`. It should take one in and write a new one out. To disk, because the only storage location for a `cxapi.CloudFormationStackArtifact` is in a file on disk (which is already... meh).

The stored state for `xxx.expected.json` is not a StackArtifact, it's just a template. So we're going to have to hydrate one to a `CloudFormationStackArtifact` in order to run it through the normalizer, which means we're stuck doing the "regex searching" there, so that we can add assets to the metadata so that they can be recognized by the normalizer.

Why not write the normalized template to disk? We could. The user experience would just be shittier: when you run into a failing test, you can't just add `pragma:ignore-assets`, you still have to run the tests again, kinda defeating the point. Also, we'd have to test for and add the normalization code in 2 places: both in `cdk-integ` as well as `cdk-integ-assert`.